src/Aspire.Hosting.AWS/Provisioning/Provisioners/AWSProvisioner.cs

```diff
@@ -0,0 +1,167 @@
+        };
+        // parent -> children lookup
+        var parentChildLookup = appModel.Resources.OfType<IResourceWithParent>()
+            .Select(x => (Child: x, Root: SelectParentAWSResource(x.Parent)))
+            .Where(x => x.Root is not null)
+            .ToLookup(x => x.Root, x => x.Child);
+
+        // Sets the state of the resource and all of its children
+        async Task UpdateStateAsync(IAWSResource resource, Func<CustomResourceSnapshot, CustomResourceSnapshot> stateFactory)
+        {
+            await notificationService.PublishUpdateAsync(resource, stateFactory).ConfigureAwait(false);
+            foreach (var child in parentChildLookup[resource])
+            {
+                await notificationService.PublishUpdateAsync(child, stateFactory).ConfigureAwait(false);
+            }
+        }
+        // After the resource is provisioned, set its state
+        async Task AfterProvisionAsync(IAWSResource resource)
+        {
+            try
+            {
+                await resource.ProvisioningTaskCompletionSource!.Task.ConfigureAwait(false);
+
+                await UpdateStateAsync(resource, s => s with
+                {
+                    State = new ResourceStateSnapshot("Running", KnownResourceStateStyles.Success)
+                }).ConfigureAwait(false);
+            }
+            catch (Exception)
+            {
+                await UpdateStateAsync(resource, s => s with
+                {
+                    State = new ResourceStateSnapshot("Failed to Provision", KnownResourceStateStyles.Error)
+                }).ConfigureAwait(false);
+            }
+        }
+        // Mark all resources as starting
+        foreach (var r in awsResources)
+        {
+            r.ProvisioningTaskCompletionSource = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            await UpdateStateAsync(r, s => s with
+            {
+                State = new ResourceStateSnapshot("Starting", KnownResourceStateStyles.Info)
+            }).ConfigureAwait(false);
+
+            // After the resource is provisioned, set its state
+            _ = AfterProvisionAsync(r);
+        }
+
+        // This is fully async, so we can just fire and forget
+        _ = Task.Run(() => ProvisionAWSResources(awsResources, cancellationToken), cancellationToken);
+    }
+
+    private async Task ProvisionAWSResources(IList<IAWSResource> awsResources, CancellationToken cancellationToken)
+    {
+        var tasks = new List<Task>();
+
+        foreach (var resource in awsResources)
+        {
+            tasks.Add(ProcessResourceAsync(resource, cancellationToken));
+        }
+
+        var task = Task.WhenAll(tasks);
+
+        // Suppress throwing so that we can save the user secrets even if the task fails
+        await task.ConfigureAwait(ConfigureAwaitOptions.SuppressThrowing);
+
+        // Set the completion source for all resources
+        foreach (var resource in awsResources)
```

Comment:
Is this done here as a defensive measure? I'm assuming the `ProcessResourceAsync` should have taken care of this. Having this here threw me off thinking we were not returning the task completion source wasn't done till all AWS provisioning was done.

Reply:
Will remove this. This is more important for the Azure piece.